### ISSUE TITLE
Problem connecting to PostgreSQL via socket

### ISSUE THREAD

**Issue by michalbachowski** [grafana/grafana#2553]
Hi,

I could not connect Grafana to PostgreSQL via local socket (instead of TCP/IP). Such way of connecting to PostgreSQL is supported in [lib pq](https://github.com/lib/pq/blob/b269bd035a727d6c1081f76e7a239a1b00674c40/conn.go#L282) thus is supported in ``xorm``.

But Grafana complains:

```bash
2015/08/19 09:01:50 [log.go:75 Fatal()] [E] fail to initialize orm engine: Sqlstore::Migration failed err: pq: password authentication failed for user "XXX"
```
And PostgreSQL logs:
```bash
2015-08-19 09:03:48.070 UTC 15186 [unknown]  [unknown] LOG:  connection received: host=127.0.0.1 port=51881
2015-08-19 09:03:48.071 UTC 15186 XXX 127.0.0.1 YYY FATAL:  password authentication failed for user "XXX"
```
What I expect from PostgreSQL is something like:
```bash
2015-08-19 09:03:29.916 UTC 15114 [unknown]  [unknown] LOG:  connection received: host=[local]
2015-08-19 09:03:29.917 UTC 15114 XXX [local] YYY LOG:  connection authorized: user=XXX database=YYY
```

I'm not 100% sure, but the ``host=[local]`` suggest connection to local socket while ``host=127.0.0.1`` suggest connection attempt to ``TCP/IP``. Any idea how to overcome the issue would be appreciated.

Best regards

Michał Bachowski

**Comment by torkelo** [comment]
whats your grafana.ini settings?

**Comment by michalbachowski** [comment]
default.ini
```ini
[database]
# Either "mysql", "postgres" or "sqlite3", it's your choice
type = sqlite3
host = 127.0.0.1:3306
name = grafana
user = root
password =

# For "postgres" only, either "disable", "require" or "verify-full"
ssl_mode = disable

# For "sqlite3" only, path relative to data_path setting
path = grafana.db
```
custom.ini
```ini
[database]
password=
ssl_mode=disable
name=YYY
user=XXX
type=postgres
host=/tmp
port=
path=graphana.db
```
(I keep path to be able to quickly switch between ``sqlite3`` and ``postgres`` backend).

**Comment by torkelo** [comment]
host setting is /tmp, that does not look right, host should be the ip and port

try
```
host=local:51881
```

I do not know what you should enter in host and port for local socket to work :(

**Comment by michalbachowski** [comment]
Please look at aftermentioned [lib pq](https://github.com/lib/pq/blob/b269bd035a727d6c1081f76e7a239a1b00674c40/conn.go#L282) source. If ``host`` begins with ``/`` it is considered unix socket - that's why I set it that way.

**Comment by torkelo** [comment]
@michalbachowski yes, but it stlil must be a valid socket path

**Comment by michalbachowski** [comment]
``/tmp`` is valid path as ``libpq`` will append socket name to it:

```go
if strings.HasPrefix(host, "/") {
	sockPath := path.Join(host, ".s.PGSQL."+o.Get("port"))
	return "unix", sockPath
}
```

**Comment by torkelo** [comment]
@michalbachowski then it should work, did you restart grafana-server after config change?

**Comment by michalbachowski** [comment]
tl;dr due to [issue in xorm](https://github.com/go-xorm/xorm/issues/285) in case of empty password connection config is incorrectly parsed. The workaround is to set password to anything (as it will be ignored), eg. "-".

Prerequisites: You have to widely open local socket connection but close for TCP/IP interfaces:
```ini
# "local" is for Unix domain socket connections only                                                                                                                                                              
local   all             all                                     trust                                                                                                                                             
# IPv4 local connections:                                                                                                                                                                                         
host    all             all             127.0.0.1/32            reject                                                                                                                                            
# IPv6 local connections:                                                                                                                                                                                         
host    all             all             ::1/128                 reject
```
In such case connecting to PostgreSQL via host should be rejected:
```bash
$psql -hlocalhost
psql: could not connect to server: Connection refused
	Is the server running on host "localhost" (::1) and accepting
	TCP/IP connections on port 5432?
FATAL:  pg_hba.conf rejects connection for host "127.0.0.1", user "XXX", database "YYY", SSL off
```
And log entry will be written by PostgreSQL (provided logging is enabled):
```
LOG:  connection received: host=localhost port=58636
FATAL:  pg_hba.conf rejects connection for host "127.0.0.1", user "mbachows", database "mbachows", SSL off
```
But connecting via socket should work (`/tmp` is default socket location):
```bash
$psql -h/tmp
psql -h/tmp
psql [...]
XXX=#
```
and log entry should appear:
```
LOG:  connection received: host=[local]
LOG:  connection authorized: user=mbachows database=mbachows
```

Given above PostgreSQL configuration the belov Grafana config should cause Grafan to fail connecting PostgreSQL:
```ini
[database]
password=
ssl_mode=disable
name=YYY
user=XXX
type=postgres
host=/tmp
port=
path=graphana.db
```
and Grafana fail:
```bash
$bin/grafana-server
[...]
2015/08/25 11:29:57 [log.go:75 Fatal()] [E] fail to initialize orm engine: Sqlstore::Migration failed err: pq: pg_hba.conf rejects connection for host "127.0.0.1", user "XXX", database "XXX", SSL off
```

This is because xorm incorrectly parses connection config:
```user=XXX password= host=/tmp port=5432 dbname=XXX sslmode=disable```

becomes:
```map[password:host=/tmp dbname:XXX sslmode:disable host:localhost port:5432 extra_float_digits:2 user:XXX```

The workaround is simple - just put anything (at least one char) as password:
```user=XXX password=- host=/tmp port=5432 dbname=XXX sslmode=disable```

which ends up being parsed to:
```map[password:- host:/tmp dbname:XXX sslmode:disable host:localhost port:5432 extra_float_digits:2 user:XXX```

And Grafana will happily connects to PostgreSQL local socket:
```bash
$bin/grafana-server
[...]
2015/08/25 11:44:45 [I] Migrator: Starting DB migration
2015/08/25 11:44:45 [I] Listen: http://0.0.0.0:3000
```

**Comment by lunny** [comment]
If your passwd is empty, you should use single quote not let it empty. 
`user=XXX password='' host=/tmp port=5432 dbname=XXX sslmode=disable`
Please ref [http://godoc.org/github.com/lib/pq](http://godoc.org/github.com/lib/pq)

**Comment by michalbachowski** [comment]
Indeed this solution works. Thank you for your help.

Best regards

Michał Bachowski

**Comment by mlara4** [comment]
couldn't set ssqlmode\r\nParameter name sslmode

**Comment by daniellee** [comment]
@mlara4 is that a question? Not sure I understand.

**Comment by mlara4** [comment]
Nevermind, problem solved already. Thank you


Sent from my T-Mobile 4G LTE Device
-------- Original message --------From: Daniel Lee <notifications@github.com> Date: 3/23/17  2:20 AM  (GMT-08:00) To: grafana/grafana <grafana@noreply.github.com> Cc: mlara4 <marklara@cox.net>, Mention <mention@noreply.github.com> Subject: Re: [grafana/grafana] Problem connecting to PostgreSQL via socket
  (#2553) 
@mlara4 is that a question? Not sure I understand.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/grafana/grafana","title":"grafana/grafana","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/grafana/grafana"}},"updates":{"snippets":[{"icon":"PERSON","message":"@daniellee in #2553: @mlara4 is that a question? Not sure I understand."}],"action":{"name":"View Issue","url":"https://github.com/grafana/grafana/issues/2553#issuecomment-288660647"}}}
